### ISSUE TITLE
Putting listener rules in the middle of the list causes erratic terraform behavior

### ISSUE THREAD

**Issue by vladistan** [terraform-aws-modules/terraform-aws-alb#289]
## Description

When adding a new rule to the middle of https_listeners array after creating the stack, Terraform fails to update the listener rule with a new ID, causing duplicate rules with the same ID. 

Steps to Reproduce:

* Apply the  attached Terraform plan containing the initial listener rules. (You need to provide your own certificate ARN)
* Uncomment the existing rule in the middle of the http_tcp_listeners or https_listeners array, marked as such
* Apply the updated plan.
* Notice the duplicate rule with the same ID is created.
* Attempt to apply the plan again without any changes to the Terraform code and notice Terraform overwriting existing rule and never converging 

Expected Results:

When adding a new rule to the middle of the http_tcp_listeners or https_listeners array, Terraform should update the listener rule with a new ID, rather than creating a duplicate rule with the same ID.

- [X] ✋ I have searched the open/closed issues and my issue is not listed.

## Versions

- Module version [Required]: 8.6.0

- Terraform version:
1.3.8
1.4.4

- Provider version(s):

registry.terraform.io/hashicorp/aws v4.62.0

## Reproduction Code [Required]

```
provider "aws" {
  region = local.region
}

data "aws_availability_zones" "available" {}

locals {
  name   = "ex-${basename(path.cwd)}"
  region = "us-east-1"

  vpc_cidr = "10.0.0.0/20"
  azs      = slice(data.aws_availability_zones.available.names, 0, 2)

}

##################################################################
# Application Load Balancer
##################################################################

module "alb" {
  source  = "terraform-aws-modules/alb/aws"
  version = "~> 8.6"

  load_balancer_type = "application"
  internal           = true

  name    = "my-alb-int"
  vpc_id  = module.vpc.vpc_id
  subnets = module.vpc.private_subnets

  security_groups = [module.alb_sg.this_security_group_id]

  https_listeners = [
    {
      port            = 443
      protocol        = "HTTPS"
      certificate_arn = var.certificate
      action_type     = "fixed-response"
      fixed_response = {
        content_type = "text/plain"
        status_code  = "200"
        message_body = "Nothing to see here move along"
      }
    },
  ]

  https_listener_rules = [
    {
      https_listener_index = 0
      priority             = 100

      actions = [{
        type               = "forward"
        target_group_index = 0
      }]

      conditions = [{ host_headers = ["my-host-header.example.com"] }]
    },
    # UNCOMMENT THIS RULE AFTER APPLYING THE PLAN FOR THE FIRST TIME
    # {
    #   https_listener_index = 0
    #   priority             = 102

    #   actions = [{
    #     type               = "forward"
    #     target_group_index = 0
    #   }]

    #   conditions = [{ host_headers = ["my-host-header2.example.com"] }]
    # },
    {
      https_listener_index = 0
      priority             = 114

      actions = [{
        type               = "forward"
        target_group_index = 0
      }]

      conditions = [{ host_headers = ["my-host-header3.example.com"] }]
    },
  ]

  target_groups = [
    {
      name_prefix          = "my-tg"
      backend_protocol     = "HTTPS"
      backend_port         = 443
      target_type          = "ip"
      deregistration_delay = 60
      health_check = {
        path                = "/"
        port                = "traffic-port"
        protocol            = "HTTPS"
        healthy_threshold   = 2
        unhealthy_threshold = 2
        timeout             = 5
        interval            = 30
      }
    }
  ]

}




################################################################################
# Supporting resources
################################################################################

module "vpc" {
  source  = "terraform-aws-modules/vpc/aws"
  version = "~> 3.0"

  name = local.name
  cidr = local.vpc_cidr

  azs             = local.azs
  private_subnets = [for k, v in local.azs : cidrsubnet(local.vpc_cidr, 4, k)]
  public_subnets  = [for k, v in local.azs : cidrsubnet(local.vpc_cidr, 8, k + 48)]

  enable_nat_gateway   = true
  single_nat_gateway   = true
  enable_dns_hostnames = true


}

module "alb_sg" {

  source  = "terraform-aws-modules/security-group/aws"
  version = "3.0.0"

  name        = "alb-sg"
  description = "Security group for the ALB"
  vpc_id      = module.vpc.vpc_id

}


variable "certificate" {
  default = "INSERT YOUR CERTIFICATE ARN HERE"
  description = "Provide the certificate ARN"
}

```

No

Yes

1.  Set the value of the 'certificate' variable to your own certificate
2.  Apply the terraform plan to create initial stack,  it will succeed
3.  Uncomment the lines containing the rule in the middle of the array

## Expected behavior

A rule added to the listener when the stack is applied with it's own unique ID

## Actual behavior

The new rule has the same ID as the rule that was in that position,  this causes each run of terraform to overwrite the rule definition either with the previous rule or the one you just added,  the stack never converges

**Comment by vladistan** [comment]
This could be related to #278

**Comment by LeandroMichelson** [comment]
> Isso pode estar relacionado com [# 278](https://github.com/terraform-aws-modules/terraform-aws-alb/issues/278)

This issue has been automatically closed!

However, I think this feature benefits many scenarios.

**Comment by bryantbiggs** [comment]
we are working on this refactor - please be patient as it will take some time to develop, test, and validate plus provide documentation to help users migrate to this new approach

**Comment by odira1** [comment]
We will be waiting patiently. Let me know if you need some help!

**Comment by darkpandarts** [comment]
I think i'm getting hit by this issue as well. Any news?

**Comment by darkpandarts** [comment]
bump

**Comment by bryantbiggs** [comment]
you can track WIP here if interested https://github.com/terraform-aws-modules/terraform-aws-alb/compare/master...bryantbiggs:refactor/for-each?expand=1

**Comment by darkpandarts** [comment]
many thanks!

**Comment by mdimarino** [comment]
Having more than one http_tcp_listeners or https_tcp_listeners makes things worse. All listerners rules creates a big list very dificult to manage.

**Comment by antonbabenko** [comment]
This issue has been resolved in version 9.0.0 :tada:
